proto/control-api/src/control/mod.rs

```diff
@@ -0,0 +1,478 @@
+///
+/// [`ControlApi`]: Api
+#[derive(Clone, Debug)]
+pub struct ErrorResponse {
+    /// [`ErrorCode`] which will be returned with code and message.
+    pub error_code: ErrorCode,
+
+    /// [`Element`] `ID` where some error happened. May be empty.
+    pub element_id: Option<String>,
+
+    /// All [`ErrorCode`]s have [`Display`] implementation. And this
+    /// implementation will be used if this field is [`None`]. But
+    /// sometimes we want to add some error explanation. Then we set this
+    /// field to [`Some`] and this text will be added to
+    /// [`Display`] implementation's text.
+    ///
+    /// By default this field should be [`None`].
+    ///
+    /// For providing error explanation use [`ErrorResponse::with_explanation`]
+    /// method.
+    ///
+    /// [`Display`]: std::fmt::Display
+    pub explanation: Option<String>,
+}
+
+impl ErrorResponse {
+    /// New [`ErrorResponse`] with [`ErrorCode`] and [`Element`] `ID`.
+    pub fn new<T: ToString>(error_code: ErrorCode, element_id: &T) -> Self {
+        Self {
+            error_code,
+            element_id: Some(element_id.to_string()),
+            explanation: None,
+        }
+    }
+
+    /// New [`ErrorResponse`] only with [`ErrorCode`].
+    #[must_use]
+    pub const fn without_id(error_code: ErrorCode) -> Self {
+        Self {
+            error_code,
+            element_id: None,
+            explanation: None,
+        }
+    }
+
+    /// [`ErrorResponse`] for all unexpected errors.
+    ///
+    /// Provide unexpected `Error` to this function.
+    /// This error will be printed with [`Display`] implementation
+    /// of provided `Error` as error explanation.
+    ///
+    /// [`Display`]: std::fmt::Display
+    pub fn unexpected<B: ToString>(unknown_error: &B) -> Self {
+        Self {
+            error_code: ErrorCode::UnexpectedError,
+            explanation: Some(unknown_error.to_string()),
+            element_id: None,
+        }
+    }
+
+    /// [`ErrorResponse`] with some additional info.
+    ///
+    /// With this method you can add additional text to error message of
+    /// [`ErrorCode`].
+    #[must_use]
+    pub const fn with_explanation(
+        error_code: ErrorCode,
+        explanation: String,
+        id: Option<String>,
+    ) -> Self {
+        Self {
+            error_code,
+            explanation: Some(explanation),
+            element_id: id,
+        }
+    }
+}
+
+/// [Medea]'s [`ControlApi`] errors.
+///
+/// [`ControlApi`]: Api
+/// [Medea]: https://git.instrumentisto.com/streaming/medea
+#[derive(Clone, Copy, Debug, Display)]
+#[repr(u16)]
+pub enum ErrorCode {
+    /// Unimplemented API call.
+    ///
+    /// This code should be with additional text which explains what
+    /// exactly unimplemented (you can do it with
+    /// [`ErrorResponse::with_explanation`] function).
+    ///
+    /// Code: __1000__.
+    #[display(fmt = "Unimplemented API call.")]
+    UnimplementedCall = 1000,
```

Comment:
I don't like the idea of hardcoding all the error codes in the protocol. This is quite unpractical, as any time we adjust the implementation - we should tune up the protocol as well.

It's better to keep the error structure and so its codes as an implementation detail, allowing implementors fulfill their needs here.

Let's just abstract over errors in the traits via `type Error` associative type. And once somebody implements it, it may provide its own error codes via `const`s or `enum`, or even go without codes and `match` the returned error directly.